lib/puppet/type/tcp_conn_validator.rb

```diff
@@ -0,0 +1,44 @@
+Puppet::Type.newtype(:tcp_conn_validator) do
+
+  @doc = "Verify that a TCP connection can be successfully established between a node
+          and the expected  server.  Its primary use is as a precondition to
+          prevent configuration changes from being applied if the server cannot be
+          reached, but it could potentially be used for other purposes such as monitoring."
+
+  ensurable do
```

Comment:
Since this resource doesn't really have an actual "state" I don't think it should be ensurable - having something like a "reachable => true/false" seems like a better approach to me.